### DISCUSSION TITLE
CreatePermission 403 Forbidden

### DISCUSSION THREAD

**Discussion by lherman-cs** [mycrl/turn-rs#104] [Q&A]
This project is awesome! I had a nice time reading the codebase and particularly enjoyed the separation of `turn` vs `turn-server.

**Describe the bug**
I was able to receive the reflexive candidate from the TURN service. However, when my client tried to create a binding or CreatePermission request, The turn server responded with 403 Forbidden.

**To Reproduce**
My TURN setup currently is the following:

1. Compile from the latest main: https://github.com/mycrl/turn-rs/commit/f113b2b42bbc926e715c6bac36b3dead1a04482e
2. Configure so the bind port listens to `0.0.0.0:3478` and the external IP is set to `<external>:3478`. The transport is set to UDP.
3. Static auth config is `demo=demo`.
4. The server ran without any containerization, just a barebone binary. The instance also has direct access to an external IP.
5. Use either https://webrtc.github.io/samples/src/content/peerconnection/trickle-ice/ or https://github.com/pion/turn/tree/master/examples/turn-client/udp to run a ping test.

**Expected behavior**
TURN server to allow the allocation and let the media flow.

**Additional context**
From tracing the codebase, these lines caused the issue, https://github.com/mycrl/turn-rs/blob/main/turn/src/processor/create_permission.rs#L93-L99.

I didn't see anything in RFC that defines this behavior as a MUST. So, I'm expecting that there are some specific reasons for this implementation.

**Regarding `ip_is_local`:**

`ctx` contains my server's external IPs, whereas `peer` contains the reflexive IP. But, with `!ip_is_local` (as the name implies), the TURN server will always reject external IPs.

**Then, I tried commenting out `ip_is_local` check:**

I found that the `bind_port` look ups for the peer's port whereas the allocation used the server's port instead, https://github.com/mycrl/turn-rs/blob/main/turn/src/processor/allocate.rs#L88-L91.

Thus, `bind_port` also couldn't find anything and ended with another 403 Forbidden.

**Comment by mycrl** [comment]
Thank you, I am happy with how my program has helped you.

First of all, I'm not complaining that your proposed problem doesn't exist or is boring, and I hope my tone doesn't bother you. This project has been pretty much tried and tested for standard stun and turn session processes, including having been used in many business systems for a long time, so there shouldn't be too many problems with basic business turn-rs.
Next, I'll walk you through troubleshooting the problem step-by-step.

First of all, the `ip_is_local` part that you brought up in your question, the meaning of this check is that if an illegitimate turn request is received, such as requesting an address that doesn't belong to the current turn server, it should be rejected, which really doesn't belong to the RFC, but this is a defensive process, and turn servers are supposed to be rejecting illegitimate requests, since It is not possible to create permissions or bind peers for requests that do not belong to the current turn server.

First, do you have your external address set to an external address, but your connectivity test is using a local address such as `127.0.0.1`, which would cause turn-rs to reject your request due to an external mismatch. Because in this case, the turn client will most likely also set the external address to `127.0.0.1` when constructing the turn request, so `ip_is_local` won't get through here, I'm not sure if other implementations of turn servers allow for this, but it seems to be reasonable for me to do it this way. I'm guessing that's most likely the case with your problem.

Secondly, you can try to run turn-rs in a local environment with both bind and external configured to `127.0.0.1` and test if it passes, if it passes, then the problem is the situation I have already mentioned above.

For example, when your `external` address is set to `192.168.1.100`, then the turn client should need to use `turn:192.168.1.100:3478` when communicating with the turn server.

**Comment by lherman-cs** [comment]
Thanks for the response. Knowing the project has been tried and tested in production is reassuring. I've probably made a mistake with my configuration or environment. 

The reject logic reasoning makes sense to me. Yes, I set the external address to the public IP.  This is the result from running both client and servers locally with externalIP set to 127.0.0.1. The request still fails on Forbidden 403.

```toml
[turn]
realm = "turn.lukas-coding.us"

[[turn.interfaces]]
transport = "udp"
bind = "127.0.0.1:3478"
external = "127.0.0.1:3478" # with external exposure, I replaced "127.0.0.1" with my "<externalIP>"

[log]
level = "info"

[api]
bind = "127.0.0.1:3001"

[auth]
[auth.static_credentials]
demo = "demo"
```

<img width="1110" alt="Screenshot 2024-11-18 at 10 09 20 AM" src="https://github.com/user-attachments/assets/67f2f407-bfba-4b0e-97fd-0a46a5e3e3ea">

<img width="834" alt="Screenshot 2024-11-18 at 10 09 39 AM" src="https://github.com/user-attachments/assets/faf3c058-5cca-43b7-963f-b912ea8bbbde">


The TURN ping test consists of the following:

1. TURN allocates for the client to learn about its relay address:port
2. Send a binding request to get the client's externalIP:port.
3. Use the allocated relay address:port (from step 1) to send a ping to the client's externalIP:port (from step 2).

On a side note, running https://webrtc.github.io/samples/src/content/peerconnection/trickle-ice/ results in no candidates. But, I see allocations happening on the server side without errors.

I have also replicated a similar setup with Coturn. Both client tests against coturn resulted in candidates.

What tools would you recommend in testing the turn server?

**Comment by mycrl** [comment]
I tested it locally and didn't check that there was a problem.

I tested using trickle ice as well, with the turn server running locally and the config file using the example config file from the project, and didn't change anything else except uncommenting `user1=test`.

![{459ABBC4-C800-4DFF-A799-B95DB0B4157E}](https://github.com/user-attachments/assets/804f3fdc-f214-490a-8336-25b523f3612d)

This is my test result, which has successfully assigned the address and created the permissions

![{58218D6E-2AC0-4B67-819E-3A313C678C28}](https://github.com/user-attachments/assets/1e1af9f9-0064-40cb-be6d-0996a0ae68d2)

![{0AAFD488-3B6F-45C7-A753-40E4343BE741}](https://github.com/user-attachments/assets/fcd8ca06-06d4-405e-85bf-b46ddbd18724)

**Comment by mycrl** [comment]
I'll test it again with https://github.com/firefart/stunner, and let me know if you have more info on it as well, because I recently posted a new version that was not in the production environment with a few changes, which theoretically shouldn't be a problem, but maybe I changed something to cause a problem with the new version .

**Comment by mycrl** [comment]
@lherman-cs I simply wrote a webrtc test page to force the use of turn forwards to create rtc, I tested it and found no problems.
Can you provide your wireshark dump file if it's convenient? If you don't want your net information to be compromised, you can send it to my private email: lepidodendraceae@gmail.com

**Comment by lherman-cs** [comment]
Interesting. I found that the Webrtc trickle ice test page works with Chrome, not Firefox. I'll check stunner and get the wireshark dump too.

## Chrome

<img width="907" alt="Screenshot 2024-11-19 at 9 02 13 AM" src="https://github.com/user-attachments/assets/b33a087b-32c8-4b5b-9efd-5f804958f39b">

<img width="1114" alt="Screenshot 2024-11-19 at 9 00 43 AM" src="https://github.com/user-attachments/assets/85d69550-5223-4a3e-8474-ad56048d09da">

## Firefox

<img width="1081" alt="Screenshot 2024-11-19 at 9 02 23 AM" src="https://github.com/user-attachments/assets/d49ff281-cf7d-4a91-85de-3d1494ce4386">


<img width="955" alt="Screenshot 2024-11-19 at 9 00 53 AM" src="https://github.com/user-attachments/assets/66792037-2696-40af-9bba-8386d8a66f86">

**Reply by lherman-cs** [reply]
I found the main problem in my test that caused CreatePermission 403 Forbidden. In my test using Pion client, the client sets the `XOR-PEER-ADDRESS` to the `XOR-MAPPED-ADDRESS` as opposed to `XOR-RELAYED-ADDRESS`. Since turn-rs uses the allocated address for lookups, this ends up being forbidden because the `XOR-MAPPED-ADDRESS` doesn't exist.


Looking at the [RFC5766](https://www.ietf.org/rfc/rfc5766.txt), the `XOR-PEER-ADDRESS` is indeed ambiguous. Tbh, this can be either `XOR-RELAYED-ADDRESS` or `XOR-MAPPED-ADDRESS`. So, not sure if it makes sense to allow this address too.

```
14.3.  XOR-PEER-ADDRESS

   The XOR-PEER-ADDRESS specifies the address and port of the peer as
   seen from the TURN server.  (For example, the peer's server-reflexive
   transport address if the peer is behind a NAT.)  It is encoded in the
   same way as XOR-MAPPED-ADDRESS [RFC5389].
```

For the Firefox issue, it remains an issue. But, I think there is a different reason. I'll create a different discussion for it.

Here are my Wireshark files: https://drive.google.com/drive/folders/1TYYIHKnsZrVXRiuNSOaT6xC_cy8n9_GM?usp=sharing

There are a bunch of them. But, the most interesting ones are `chrome-real-w-relay` (test with a real stream) and `pion-w-ping` (the pion client test I originally used).

**Reply by mycrl** [reply]
@lherman-cs The flow of a TURN session is like this:

1, first of all, the client requests to allocate an address, at this time the client sends an `Allocate` request, how the request is no problem, the server responds with a `XOR-RELAYED-ADDRESS`, the address is the address and port number that the server allocates to the client, the ip address is the server's external ip address, and the port number is the server's random allocation of a port number randomly assigned by the server. The `XOR-RELAYED-ADDRESS` is the more important one, theoretically the client only needs to pay attention to this address, as for the `XOR-MAPPED-ADDRESS` address, it is for the server, the current client's external address, which is the address of the sender of the UDP packet that the server receives, and this address is replied back to the client to let the the client can determine its own NAT type blah blah blah.

2, this time into the `CreatePermission` process, the client first need to inform the server you need to apply for permission to the other side of the forwarding address, this address is the above `XOR-RELAYED-ADDRESS` assigned address, note that this address is the other side of the address, not their own address is A and B each other `CreatePermission`, that is, A and B each other. CreatePermission`, also known as `XOR-PEER-ADDRESS`, this is the address that the server assigns to the other party, and the client tells the server, I need to create permission and pair with the other party.

I looked at the wireshark dump for `pion` and in the `CreatePermission` phase, the client creates the permission for the `XOR-RELAYED-ADDRESS` that the server assigns to itself, which is theoretically wrong, but does the RFC allow this behavior? I don't know.

Of course, I can be compatible with this behavior if creating permissions for myself and the server does nothing is fine and returns a creation success. I'm under the impression that this doesn't seem to be a behavior that the RFC explicitly says needs to be supported (or maybe there's an error in my implementation and it's allowed).

At least, turn-rs doesn't allow this behavior, since it seems pointless to create permissions for yourself; the client itself has permissions to the forwarding address that the server assigns to itself, and the client doesn't forward it to itself.

**Reply by lherman-cs** [reply]
@mycrl thanks for the explanation. It's aligned with my understanding as well. Although I would argue that `XOR-PEER-ADDRESS` is loosely defined in RFC, the usage here is merely a unique ID representing the peer. The main confusion seems to stem from whether a peer is uniquely identified by `XOR-MAPPED-ADDRESS` or `XOR-RELAYED-ADDRESS`. The `XOR-MAPPED-ADDRESS` opened on the closest NAT to turn-rs should be unique to the peer. 

Moreover, turn-rs technically has both addresses that uniquely map to the peer. I agree that the concept of 'creating permission for yourself' feels semantically odd, but, I think it helps to think of this as a reverse lookup with an index to a database as an analogy.

As we can't control how third-party WebRTC implementation (server side such as webrtc-rs, pion, etc.), we're also risking incompatibility with these clients. With more use cases like SFU, MCU, and potentially AI bots in the future, this can be prohibitive for production use.

**Reply by mycrl** [reply]
Good point, I'll try compatibility and merge it into v3.2.

**Reply by mycrl** [reply]
@lherman-cs Can you provide your test program? I would like to fix this problem.

**Reply by lherman-cs** [reply]
@mycrl https://github.com/pion/turn/tree/master/examples/turn-client/udp this is what I used. 

Not sure how familiar you are with Go, you just need `go build` in the directory. It'll place the binary in the same folder.

**Reply by mycrl** [reply]
Thanks, it's ok, I'm ok with go language, the go language is fairly simple...

**Reply by mycrl** [reply]
@lherman-cs I released a beta version of 3.2 for you to try.

**Reply by lherman-cs** [reply]
The Forbidden error still exists. `ip_is_local` (https://github.com/mycrl/turn-rs/pull/106/files#diff-244d1430cca1ece39f98009197f22cc02164be5ac028b963dbe58038f1e48b61R93) still rejected. Since `external_ip` must be set to the turn server's IP, the peer's IP doesn't match. 

Even after removing the check, the turn test passed the CreatePermission request but no data came through. Unfortunately, I don't have much bandwidth to dig deeper into this issue.

This problem is not as important as the Firefox issue for me. So, please feel free to put this issue as a backlog.

**Reply by mycrl** [reply]
external must be the external ip of the turn server this is required, even in the RFC this is explicitly checked.

In the RFC, the port of the peer address can be ignored, but the ip part must be an address that the turn server can be assigned.

As a typical example, if in the configuration file, your external is 192.168.1.100, which is the external ip address of the turn server, and you make a turn request directly from this machine, such as `turn:127.0.0.1:3478`, then this case will be rejected for sure. Because of the mismatch, normally turn servers are not allowed to be used in this way. If you use `turn:192.168.1.100:3478` to initiate the request there is no problem.

I tested the beta version using both pion and webrtc-rs examples and found no problems.
And I also tested real webrtc scenarios using chrome and firefox, creating 4 rtc peer connectin to each other at the same time for half an hour to an hour of real world testing, and no problems.

I think your problem may be due to your complex network topology or problems with the turn client implementation.

You can provide more information if you are interested.

**Reply by mycrl** [reply]
Also, are you using v3.2.0-beta or v3.2.0-beta.1 for testing? Theoretically, these two versions don't do much of anything to limit permission creation requests anymore, they normally create permissions successfully and don't make any state changes to the permission creation requests, so if you can't successfully do the subsequent steps even after commenting out the ip check, that's not the problem.

I'll keep that principle, I'm happy to help you out, and if you can provide a wireshark dump all the better.

ps: https://github.com/mycrl/turn-rs/blob/main/turn/src/operation/create_permission.rs#L78 Now this part doesn't actually change any state, so the fact that you can't continue the process after commenting it out has nothing to do with this part really. It has nothing to do with this section.

**Comment by mycrl** [comment]
After such a long time, I bring rather depressing news.

Having recently gone through the details of the RFC and the coturn implementation from scratch and re-implemented the turn-rs session handling, I've found that coturn is more forgiving and implements the details in the RFC perfectly.

I tried to get turn-rs to implement behaviour consistent with coturn, especially when creating permissions and channel bindings, but I found it very difficult to implement.

For example, coturn correctly handles the fact that the `xor peer address` is an address known to a peer, not an address known to the turn server, such as the typical IP of an intranet device, but I've been working for a long time on how to create associations where the two parties aren't related, because for turn-rs I don't know the association between the two parties! It's a challenge to forward the message correctly.

I didn't really look at the coturn implementation, as there are advantages to only allowing both parties to communicate using the address assigned by the turn server, which allows both parties to hide their identities and actual addresses.

However, turn-rs currently works in the vast majority of cases, at least the common clients work as expected, and basically the common implementations work as expected in non-intranet environments.

This is undeniably an issue, but for now it can be tentatively classified as some limitation of turn-rs rather than a bug.
